File: crates/musicata-server/web/src/lib/autoeq.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// AutoEq headphone correction presets, fetched on demand through Musicata's own API.
//
// Deliberately NOT bundled. AutoEq's own code is MIT, but its published results are computed
// from headphone measurements contributed by third parties (oratory1990, Crinacle, Rtings and
// others), each publishing under its own terms — some non-commercial. Musicata stores nothing:
// the server relays each response straight through (see crate::proxy) and a preset only ever
// comes to rest in the user's own saved profile. See NOTICE.
//
// The page cannot reach raw.githubusercontent.com itself — the CSP allows this origin only —
// so both requests go to the server, which fetches upstream. That also keeps the user's IP
// from reaching a third party the operator never configured. `path` is sent exactly as
// AutoEq's INDEX.md spells it; the server derives the filename and rejects anything that would
// escape the results/ base (see crate::autoeq).

const API = "/api/autoeq";

export interface AutoEqEntry {
  /** Headphone model, as AutoEq names it — e.g. "Sennheiser HD 600". */
  name: string;
  /** Who measured it. Several sources publish the same model, with different results. */
  source: string;
  /** Measurement rig, when the source publishes on more than one. */
  rig?: string;
  /** Path under `results/`, percent-encoded as it appears in the index. */
  path: string;
}

/**
 * Parse AutoEq's `results/INDEX.md`. Lines look like:
 *
 *     - [Sennheiser HD 600](./oratory1990/over-ear/Sennheiser%20HD%20600) by oratory1990
 *     - [1MORE Aero (ANC Off)](./DHRME/in-ear/1MORE%20Aero%20(ANC%20Off)) by DHRME
 *     - [Sennheiser HD 600](./crinacle/GRAS%2043AG-7%20over-ear/…) by crinacle on GRAS 43AG-7
 *
 * Model names contain parentheses, so the link group is matched **greedily** and anchored on
 * the trailing `) by …` — a lazy match would stop inside `(ANC Off)` and truncate the path.
 * Unparseable lines (the file's own headings and prose) are skipped.
 */
export function parseIndex(markdown: string): AutoEqEntry[] {
  const entries: AutoEqEntry[] = [];
  for (const raw of markdown.split(/\r?\n/)) {
    const m = raw.match(/^- \[(.+)\]\(\.\/(.+)\) by (.+)$/);
    if (!m) continue;
    const [, name, path, by] = m;
    const on = by.indexOf(" on ");
    entries.push({
      name,
      path,
      source: on === -1 ? by : by.slice(0, on),
      rig: on === -1 ? undefined : by.slice(on + 4),
    });
  }
  return entries;
}

/** Models whose name contains every whitespace-separated term, prefix matches first. */
export function searchIndex(entries: AutoEqEntry[], query: string, limit = 10): AutoEqEntry[] {
  const terms = query.trim().toLowerCase().split(/\s+/).filter(Boolean);
  if (terms.length === 0) return [];
  const matches = entries.filter((e) => {
    const name = e.name.toLowerCase();
    return terms.every((t) => name.includes(t));
  });
  // A search for "hd 600" should surface "Sennheiser HD 600" ahead of "Sennheiser HD 600 (foo)",
  // and both ahead of a model that merely contains the terms somewhere in the middle.
  const rank = (e: AutoEqEntry) => {
    const name = e.name.toLowerCase();
    return (name.startsWith(terms[0]) ? 0 : 1) * 1000 + name.length;
  };
  return matches.sort((a, b) => rank(a) - rank(b) || a.name.localeCompare(b.name)).slice(0, limit);
}

/**
 * Where to ask the server for an entry's ParametricEQ file. `entry.path` is already
 * percent-encoded (it comes verbatim out of INDEX.md), so encoding it again as a query value
 * is correct — the server decodes exactly once and gets the path back as AutoEq spells it.
 */
export function presetUrl(entry: AutoEqEntry): string {
  return `${API}/preset?path=${encodeURIComponent(entry.path)}`;
}

// The index is ~850 KB (~110 KB over the wire) and changes rarely, so it is fetched at most
// once per page load. The promise — not the result — is memoized so concurrent callers share
// one request; a failed fetch clears it so the next attempt retries instead of caching the error.
let pending: Promise<AutoEqEntry[]> | null = null;

export function loadIndex(): Promise<AutoEqEntry[]> {
  pending ??= fetch(`${API}/index`)
    .then((res) => {
      if (!res.ok) throw new Error(`AutoEq index: HTTP ${res.status}`);
      return res.text();
    })
    .then(parseIndex)
    .catch((err) => {
      pending = null;
      throw err;
    });
  return pending;
}

export async function fetchPreset(entry: AutoEqEntry): Promise<string> {
  const res = await fetch(presetUrl(entry));
  if (!res.ok) throw new Error(`AutoEq preset: HTTP ${res.status}`);
  return res.text();
}

File: crates/musicata-server/src/autoeq.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// Musicata — a local-first music server + web controller.
// Copyright (C) 2026 Damir Krstanović
//
// This program is free software: you can redistribute it and/or modify it under the terms of
// the GNU Affero General Public License as published by the Free Software Foundation, either
// version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License along with this
// program. If not, see <https://www.gnu.org/licenses/>.

//! URL construction for the AutoEq headphone-correction proxy.
//!
//! The browser cannot reach `raw.githubusercontent.com` (the CSP allows `'self'` only), so it
//! asks Musicata for the model index and for a chosen model's ParametricEQ file, and the server
//! fetches them (see [`crate::proxy`]). Presets are relayed, never stored — a Musicata instance
//! holds no copy of AutoEq's corpus.
//!
//! The security property this module exists to guarantee: **the client names a model, never a
//! URL.** The base is a constant, the caller supplies only a path *suffix*, and
//! [`preset_url`] rejects anything that could escape that suffix into a different host, a
//! parent directory, or a query of its own. Without that check the endpoint would be an open
//! proxy sitting inside the user's LAN.

/// AutoEq publishes its computed results here; `INDEX.md` lists every model.
const RESULTS_BASE: &str = "https://raw.githubusercontent.com/jaakkopasanen/AutoEq/master/results";

/// The model index — one line per measured headphone.
pub fn index_url() -> String {
    format!("{RESULTS_BASE}/INDEX.md")
}

/// The ParametricEQ file for `path`, which is a `results/`-relative directory exactly as
/// `INDEX.md` spells it (already percent-encoded, e.g. `oratory1990/over-ear/Sennheiser%20HD%20600`).
///
/// AutoEq names the file after the directory's last segment, so the caller supplies only the
/// directory and the filename is derived here — the client never gets to choose it.
pub fn preset_url(path: &str) -> anyhow::Result<String> {
    // Check the caller's string BEFORE normalising it. Trimming first would let `//evil/x`
    // through: the trim would eat the leading pair and the `//` test would then see nothing.
    // Each of these would let the suffix escape the fixed base: `..` climbs out of `results/`,
    // `://` and a leading `//` reroute to another host, and `?`/`#` truncate the path so the
    // derived filename is dropped. A backslash is normalised to `/` by some proxies, and
    // control characters allow request smuggling.
    for forbidden in ["..", "://", "//", "?", "#", "\\"] {
        if path.contains(forbidden) {
            anyhow::bail!("AutoEq path may not contain {forbidden:?}");
        }
    }
    if path.chars().any(|c| c.is_control()) {
        anyhow::bail!("AutoEq path may not contain control characters");
    }

    let path = path.trim_matches('/');
    if path.is_empty() {
        anyhow::bail!("empty AutoEq path");
    }

    let model = path
        .rsplit('/')
        .next()
        .filter(|segment| !segment.is_empty())
        .ok_or_else(|| anyhow::anyhow!("AutoEq path has no model segment"))?;
    Ok(format!("{RESULTS_BASE}/{path}/{model}%20ParametricEQ.txt"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn builds_the_preset_url_from_the_directory() {
        assert_eq!(
            preset_url("oratory1990/over-ear/Sennheiser%20HD%20600").unwrap(),
            format!(
                "{RESULTS_BASE}/oratory1990/over-ear/Sennheiser%20HD%20600/Sennheiser%20HD%20600%20ParametricEQ.txt"
            )
        );
    }

    #[test]
    fn tolerates_surrounding_slashes() {
        let expected = preset_url("DHRME/in-ear/1MORE%20Aero").unwrap();
        assert_eq!(preset_url("/DHRME/in-ear/1MORE%20Aero/").unwrap(), expected);
    }

    /// The whole point of the module: none of these may produce a URL.
    #[test]
    fn refuses_anything_that_escapes_the_base() {
        for path in [
            "",
            "/",
            "../../../etc/passwd",
            "a/../../b",
            "https://evil.example/x",
            "//evil.example/x",
            "model?x=1",
            "model#frag",
            "model\\..\\x",
            "model\u{0}name",
            "model\nname",
        ] {
            assert!(
                preset_url(path).is_err(),
                "should have refused {path:?} but got {:?}",
                preset_url(path)
            );
        }
    }

    /// Real AutoEq model names carry spaces (percent-encoded), parentheses and non-ASCII —
    /// the guard must not reject those.
    #[test]
    fn accepts_real_model_names() {
        for path in [
            "oratory1990/over-ear/Sennheiser%20HD%20600",
            "DHRME/in-ear/1MORE%20Aero%20(ANC%20Off)",
            "crinacle/GRAS%2043AG-7%20over-ear/Focal%20Clear",
            "rtings/over-ear/Beyerdynamic%20DT%20990%20PRO",
        ] {
            assert!(preset_url(path).is_ok(), "should have accepted {path:?}");
        }
    }

    #[test]
    fn index_is_under_the_same_base() {
        assert!(index_url().starts_with(RESULTS_BASE));
        assert!(index_url().ends_with("/INDEX.md"));
    }
}
